### ISSUE TITLE
pinned tensors need asynchronous copy routines

### ISSUE THREAD

**Issue by soumith** [torch/cutorch#216]
This came up while discussing something with NVIDIA. 

Thanks to @dominikgrewe , we have the ability to allocate pinned tensors. However,

```lua
A = cutorch.createCudaHostTensor(3);
B = torch.CudaTensor(3);
B:copy(A); — this call should be asynchronous because A is pinned memory.
```

the example above is not asynchronous because our host->device copy routines are implemented via cudaMemcpy, and not cudaMemcpyAsync:

https://github.com/torch/cutorch/blob/master/torch/generic/Tensor.c#L573
https://github.com/torch/cutorch/blob/master/lib/THC/THCTensorCopy.c#L15

We need to create a mechanism to enable async copy on the pinned versions.

Should we do this directly in :copy() itself, or should we introduce a more painful :copyAsync()? Thoughts? cc: @wickedfoo @dominikgrewe

**Comment by wickedfoo** [comment]
I'd for better or worse add copyAsync.

While :copy() is either synchronous or asynchronous with respect to the CPU
depending upon the location of arguments, it always is synchronous if
copying to or from the CPU. The fact that GPU to GPU copy is async is an
implementation detail that is not visible to the user unless they go off
the default stream, since the GPUs are effectively a black box that you
can't inspect unless you sync the GPU to CPU.

Similarly, the new :copyAsync() would be asynchronous when copying to or
from the CPU, and would map to old copy behavior for GPU to GPU copies. One
should be able to still copy synchronously to/from pinned memory, and
plugging pinned async copy into :copy() will remove the synchronous
ability. However, it should error if you're attempting an async copy
to/from CPU on non-pinned memory.

This is still weird admittedly. The only "clean" thing I think would be to
make :copy() synchronous with respect to the CPU everywhere, and
:copyAsync() async with respect to the CPU everywhere, but that would
require hunting down and changing more than half of all copy calls to
copyAsync, since otherwise GPU to GPU copy() would cudaStreamSynchronize or
event wait on the CPU to the GPU copy completion.

However the bigger problem is how one deals with the completion of the
async pinned memory copy on the CPU side? If you want to use pinned memory
as an upload buffer, what would the Torch code look like to wait for event
completion and upload again? Using device or stream synchronize for the
completion waiting kind of defeats the point of async copy from the CPU,
unless you can launch a bunch of other work once you started the copy in
the meantime before waiting. This becomes less viable if you have multiple
async buffers going at staggered intervals at the same time. I think you'd
want to think about some kind of event driven mechanism in Torch eventually.

**Comment by wickedfoo** [comment]
Regarding erroring on copyAsync for non-pinned memory: my understanding as
well is you can cudaMemcpyAsync non pinned memory to the GPU as well, but
that is slower than a sync copy, since internally the driver will copy your
memory to a small paged locked buffer and then async copy from that, only
returning to the CPU once the last chunk of your data has been copied to
that page locked buffer. It's only truly async for small data (that fits in
that buffer), and maybe not really faster than the sync copy in that
case. So copyAsync could in fact work for any tensor anywhere, but users
who don't know this detail would be tempted to start sprinkling it
everywhere, even for non-pinned memory, which will only slow down their
code.

**Comment by dominikgrewe** [comment]
Just to clarify: What is the advantage of using cudeMemcpyAsync for pinned memory? Is there any other benefit other than being able to overlap data transfer and computation when using multiple streams? For that scenario, I agreed with @wickedfoo that we should have a `copyAsync`.

**Comment by wickedfoo** [comment]
@dominikgrewe I think that's the only reason. It's only worthwhile if you are overlapping h2d data transfer with device computation, which you can only get if you go off the default stream on the device anyways.

**Comment by hughperkins** [comment]
I've been working under the assumption that I can launch tons of kernels, and copy as many buffers from host to device as I want, whilst another kernel is running.  Is that a misconception on my part?

**Comment by dominikgrewe** [comment]
With a single stream, everything will happen in order, e.g. your data transfers to/from the GPU will not overlap with computation on the GPU.

**Comment by hughperkins** [comment]
Ah, interesting :-)

**Comment by wickedfoo** [comment]
You can use two streams to do this, but your have to run your kernels off of the default stream (and the copy as well) in order to get the overlap.

If the default (NULL) stream is still involved somewhere, no matter how many other streams you are using, you won't necessarily get the result you expect.

**Comment by soumith** [comment]
fixed now via #223
